File: frontend/src/components/benchmark/content/jaccard_similarity/implementations/jaccard_similarity.php
<?php

class JaccardSimilarity {
    private array $adjacency = [];
    private array $nodes = [];

    public function addEdge(int $u, int $v): void {
        if (!isset($this->adjacency[$u])) {
            $this->adjacency[$u] = [];
        }
        if (!isset($this->adjacency[$v])) {
            $this->adjacency[$v] = [];
        }
        $this->adjacency[$u][$v] = true;
        $this->adjacency[$v][$u] = true;
        $this->nodes[$u] = true;
        $this->nodes[$v] = true;
    }

    public function buildFromEdges(array $edges): void {
        foreach ($edges as $edge) {
            $this->addEdge($edge[0], $edge[1]);
        }
    }

    public function jaccardSimilarity(int $u, int $v): float {
        if (!isset($this->adjacency[$u]) || !isset($this->adjacency[$v])) {
            return 0.0;
        }

        $neighborsU = array_keys($this->adjacency[$u]);
        $neighborsV = array_keys($this->adjacency[$v]);

        $intersection = count(array_intersect($neighborsU, $neighborsV));
        $unionSize = count($neighborsU) + count($neighborsV) - $intersection;

        if ($unionSize == 0) {
            return 0.0;
        }

        return $intersection / $unionSize;
    }

    public function scoreAllPairs(): array {
        $scores = [];
        $nodeList = array_keys($this->nodes);

        for ($i = 0; $i < count($nodeList); $i++) {
            for ($j = $i + 1; $j < count($nodeList); $j++) {
                $u = $nodeList[$i];
                $v = $nodeList[$j];
                $score = $this->jaccardSimilarity($u, $v);
                if ($score > 0) {
                    $scores[] = [
                        'pair' => [$u, $v],
                        'score' => $score
                    ];
                }
            }
        }

        usort($scores, function($a, $b) {
            return $b['score'] <=> $a['score'];
        });

        return $scores;
    }

    public function topKPredictions(int $k = 10): array {
        $scores = $this->scoreAllPairs();
        return array_slice($scores, 0, $k);
    }
}

?>

File: frontend/src/components/benchmark/content/jaccard_similarity/implementations/jaccard_similarity.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class JaccardSimilarity
{
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    public void AddEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v))
            adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        adjacency[v].Add(u);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public double JaccardSimilarity(int u, int v)
    {
        if (!adjacency.ContainsKey(u) || !adjacency.ContainsKey(v))
            return 0.0;

        var neighborsU = adjacency[u];
        var neighborsV = adjacency[v];

        int intersection = neighborsU.Intersect(neighborsV).Count();
        int unionSize = neighborsU.Union(neighborsV).Count();

        if (unionSize == 0)
            return 0.0;

        return (double)intersection / unionSize;
    }

    public List<((int, int), double)> ScoreAllPairs()
    {
        var scores = new List<((int, int), double)>();
        var nodeList = nodes.ToList();

        for (int i = 0; i < nodeList.Count; i++)
        {
            for (int j = i + 1; j < nodeList.Count; j++)
            {
                int u = nodeList[i];
                int v = nodeList[j];
                double score = JaccardSimilarity(u, v);
                if (score > 0)
                {
                    scores.Add(((u, v), score));
                }
            }
        }

        return scores.OrderByDescending(x => x.Item2).ToList();
    }

    public List<((int, int), double)> TopKPredictions(int k = 10)
    {
        return ScoreAllPairs().Take(k).ToList();
    }
}